lib/icinga/downtime.cpp

```diff
@@ -433,7 +434,11 @@ void Downtime::TriggerDowntime()
 		<< "Triggering downtime '" << GetName() << "' for checkable '" << checkable->GetName() << "'.";
 
 	if (GetTriggerTime() == 0)
-		SetTriggerTime(Utility::GetTime());
+		SetTriggerTime(triggerTime ? triggerTime : (
+			GetFixed() || !checkable->GetLastCheckResult()
+				? std::fmax(GetEntryTime(), GetStartTime())
+				: checkable->GetLastCheckResult()->GetExecutionEnd()
+		));
```

Comment:
That's a lot of ternary operator use here. Why not go for something like `if (triggerTime == 0) { triggerTime = ...; }; SetTriggerTime(triggerTime);` instead?

Also the `|| !checkable->GetLastCheckResult()` part of the condition is there only to prevent a null pointer dereference in `checkable->GetLastCheckResult()->GetExecutionEnd()` even though this should never happen for a flexible downtime that wasn't triggered by another downtime?

When suggesting the timestamp parameter, I actually had in mind to never determine the timestamp within this function but always let the caller decide what's appropriate. So when triggering from a fixed downtime check, just pass the fixed start time, when triggering due to a check result, pass the timestamp of that checkresult, and finally, when triggering from another downtime, just pass on the trigger time. This should allow the remove the need for the (hopefully) redundant null check.

Reply:
1. Is it really that hard to read?
2. Can't downtimes be triggered across checkables?
3. Tbh. I don’t want to duplicate the algo even partially, even once. Not to mention **every single TriggerDowntime()**.

Reply:
1. Harder than it needs to be IMHO.
2. Yes, but then `triggerTime` should already be set.
3. The point is that you don't need to duplicate "the algorithm" but instead, in each case where you have to give a trigger time, the choice is obvious because you know why you are triggering the downtime. I gave this a try in https://github.com/Icinga/icinga2/commit/c8e7fc836ed08020a0d90a3ca601a13f88484d41 and the only duplication is the `std::fmax` call, but that could be moved to a function if needed.

Reply:
Looks good, make a PR of it.